docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,240 @@
+| `isunion(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of the union type |
+| `isintersection(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of the intersection type |
+| `istable(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a `table` type |
+| `ismetatable(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a metatable represented as a special property of the `table` type |
+| `isfunction(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a `function` type |
+| `isclass(arg: typelib)` | `boolean` | returns true if the argument is syntactically a runtime representation of a `class` type |
+
+#### Any
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is a subtype or equal to arg in the type hierarchy |
+| `equalsto(arg: typelib)` | `boolean` | returns true if self is syntactically equal to arg in the type hierarchy |
+
+* `issubtypeof(arg: typelib)` and `equalsto(arg: typelib)` will also overload `__le` and `__eq` respectively
+    * e.g. `t1 <= t2` is equivalent to `t1:issubtypeof(t2)`
+    * e.g. `t1 == t2` is equivalent to `t1:equalsto(t2)`
+
+#### Negation
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `gettype()` | `typelib` | returns the runtime representation of the self's type being negated |
+
+#### String
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getmetatable()` | `typelib` | returns the runtime representation of self's metatable |
+
+#### StringSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `string` | returns self's value of a string singleton |
+
+#### BooleanSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `boolean` | returns self's boolean singleton value of either `true` or `false` |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `addprop(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
+| `delprop(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it (equivalent of `addprop(key, nil)`); if the key doesn't exist, nothing happens |
+| `getprops()` | `{[typelib]: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib.getstringsingleton("age") = typelib.getnumber()}`) |
```

Comment:
Can this be removed in favor of `__index`? For iterating over tables, we can provide `__iter` as well.

```luau
local foo = t["foo"]
local bar = t.bar

for k, v in t do -- k : ty, v : ty
  -- ...
end
```

Reply:
hm. I've just realized that this conflicts with `__iter` on `ty` itself (i.e. all types are iterable), which is something I wanted to see.

Then for the `rawget` example, it's easy to distribute type arguments.

```luau
type function rawget(t, p)
  local res = {}
  for _, t in t do
    for _, p in p do
      assert(t:istable())
      table.insert(res, t:getprops()[p])
    end
  end
  return ty.union(res)
end

-- "a" | "b" | "x" | "y"
type T = rawget<{ p: "a", q: "x" } | { p: "b", q: "y" }, "p" | "q">
```

Reply:
It also invites a lot of nested loops per type parameter.